### ISSUE TITLE
[Question] Is there better options for image denoising?

### ISSUE THREAD

**Issue by XCanG** [bloc97/Anime4K#104]
Right now I mostly stick with this setup, as it looks best out of 6 presets coming from readme:
```
shaders="~~/shaders/Anime4K_Denoise_Bilateral_Mode.glsl;~~/shaders/Anime4K_DarkLines_HQ.glsl;~~/shaders/Anime4K_ThinLines_HQ.glsl;~~/shaders/Anime4K_Upscale_CNN_M_x2_Deblur.glsl"
```
however I have artifacts with it in some places. I figured out, that the reason of it `Anime4K_Denoise_Bilateral_Mode.glsl`.
Here is an example of this artifact:
![image](https://user-images.githubusercontent.com/4011523/94015673-9e55da80-fdc6-11ea-98d1-c163f3209359.png)
basically it was a subtitles in opening, which become noisy and ruin picture.

If there are better alternatives, I would rather try them. I have pretty good components which should be able to handle slower algorithms.

If you need settings, here is related to render:
```
profile=gpu-hq
vo=gpu
hwdec=nvdec-copy
hwdec-codecs=all
gpu-api=vulkan

deband=yes
deband-iterations=4
deband-threshold=82
deband-range=16
deband-grain=24

scale=ewa_lanczossharp
cscale=ewa_lanczossharp

glsl-shaders="~~/shaders/Anime4K_Denoise_Bilateral_Mode.glsl;~~/shaders/Anime4K_DarkLines_HQ.glsl;~~/shaders/Anime4K_ThinLines_HQ.glsl;~~/shaders/Anime4K_Upscale_CNN_M_x2_Deblur.glsl"
```

**Comment by Jules-A** [comment]
I use Anime4K_Denoise_Bilateral_Median.glsl and change INTENSITY_SIGMA to 0.15. It's quite a bit better than mode but is heavier and needs to be used with at least Anime4K_Upscale_CNN_M_x2_Deblur to counteract the blurring. It unfortunately is nowhere near as good on text as MadVR but the only other better denoiser is the heavy which is so strong it destroys image quality on normal videos and the upscaled (Anime4K_Upscale_CNN_M_x2_Denoise) ones but they are too heavy for my system.

**Comment by XCanG** [comment]
So, `Anime4K_Upscale_CNN_M_x2_Denoise` may be better, but slow, then `Anime4K_Denoise_Bilateral_Median.glsl` alternative, but have artifacts in different place?
And how slow is MadVR? Or else how much it can destruct image? Are there any benefit from it?

**Comment by XCanG** [comment]
Hm, with `Anime4K_Upscale_CNN_M_x2_Denoise` it definitely better:
![image](https://user-images.githubusercontent.com/4011523/94019829-a7957600-fdcb-11ea-842c-66ac5cdb6247.png)
I, however can't say anything about performance impact, as statistics on `I` all zeroes on Vulkan.

**Comment by Jules-A** [comment]
> 
> 
> So, `Anime4K_Upscale_CNN_M_x2_Denoise` may be better, but slow, then `Anime4K_Denoise_Bilateral_Median.glsl` alternative, but have artifacts in different place?
> And how slow is MadVR? Or else how much it can destruct image? Are there any benefit from it?

MadVR can't be used in MPV and newer Anime4K shaders can't be used with it either since Anime4K is GLSL only now. MadVR has many different ways to configure but currently I use only use it for 480/720p, it has it's draw backs like adding a significant amount of blur to the image but it's far better than a tonne of noise and isn't very expensive either. 
I actually tested my settings and it looks like line thinning/darkening is making the text worse, rather than the denoiser.

**Comment by XCanG** [comment]
What about other Denoise files? I see several, but not sure how can I distinguish them:
Anime4K_Upscale_CNN_L_x2_Denoise.glsl
Anime4K_Upscale_CNN_M_x2_Denoise.glsl
Anime4K_Upscale_CNN_UL_x2_Denoise.glsl

**Comment by XCanG** [comment]
> I actually tested my settings and it looks like line thinning/darkening is making the text worse, rather than the denoiser.

This is true, but it work in other places, I don't see if I can replace it to something else as it work both with borders of characters and text. I just think that without AI you can't distinguish them.

**Comment by Jules-A** [comment]
> 
> 
> What about other Denoise files? I see several, but not sure how can I distinguish them:
> Anime4K_Upscale_CNN_L_x2_Denoise.glsl
> Anime4K_Upscale_CNN_M_x2_Denoise.glsl
> Anime4K_Upscale_CNN_UL_x2_Denoise.glsl

They are pretty much the same except L and UL are heavier on the system and slightly better quality.

**Comment by XCanG** [comment]
> Hm, with `Anime4K_Upscale_CNN_M_x2_Denoise` it definitely better:

Oh, wait, I just found what it not appearing in shaders list, it looks like it can't load or something.

**Comment by Jules-A** [comment]
> 
> 
> > Hm, with `Anime4K_Upscale_CNN_M_x2_Denoise` it definitely better:
> 
> Oh, wait, I just found what it not appearing in shaders list, it looks like it can't load or something.

I think maybe you forgot .glsl, I forgot to add it after copying and pasting from Windows (which doesn't include the extension).

**Comment by XCanG** [comment]
> I think maybe you forgot .glsl, I forgot to add it after copying and pasting from Windows (which doesn't include the extension).

no, I set this binds:

```
CTRL+7 no-osd change-list glsl-shaders set "~~/shaders/Anime4K_Upscale_CNN_M_x2_Denoise.glsl;~~/shaders/Anime4K_DarkLines_HQ.glsl;~~/shaders/Anime4K_ThinLines_HQ.glsl;~~/shaders/Anime4K_Upscale_CNN_M_x2_Deblur.glsl"; show-text "Anime4k: 1080p (Perceptual Quality, CNN M Denoise)"
CTRL+8 no-osd change-list glsl-shaders set "~~/shaders/Anime4K_Upscale_CNN_L_x2_Denoise.glsl;~~/shaders/Anime4K_DarkLines_HQ.glsl;~~/shaders/Anime4K_ThinLines_HQ.glsl;~~/shaders/Anime4K_Upscale_CNN_M_x2_Deblur.glsl"; show-text "Anime4k: 1080p (Perceptual Quality, CNN L Denoise)"
CTRL+9 no-osd change-list glsl-shaders set "~~/shaders/Anime4K_Upscale_CNN_UL_x2_Denoise.glsl;~~/shaders/Anime4K_DarkLines_HQ.glsl;~~/shaders/Anime4K_ThinLines_HQ.glsl;~~/shaders/Anime4K_Upscale_CNN_M_x2_Deblur.glsl"; show-text "Anime4k: 1080p (Perceptual Quality, CNN UL Denoise)"
```
All files in place (and I copy it's name, so there is no typo).

**Comment by XCanG** [comment]
There is output on all CTRL+7 to 9:
![image](https://user-images.githubusercontent.com/4011523/94022720-d5c88500-fdce-11ea-814e-83b21e2ba5dc.png)

**Comment by XCanG** [comment]
I try to switch and disable some settings, like Vulkan, however it still can't load for whatever reason... Console is also clear.

**Comment by Jules-A** [comment]
It may be because you are using 1080p monitor with 1080p content. Because of this line: 
>//!WHEN OUTPUT.w NATIVE.w / 1.200 > OUTPUT.h NATIVE.h / 1.200 > * 

It only upscales when the display resolution is 1.2x the height or (?) width of the content you are trying to watch. On my 1440p monitor, when I try watching in a smaller window with 1080p content, the shaders won't load. You can try editing that line but I'm not sure what the consequences are.

**Comment by XCanG** [comment]
Oh, I really have 1080p, but I can use DXR for make it 4k. Let me try it.

**Comment by XCanG** [comment]
Yes, now it work. And yes I don't see at this moment any artifacts at the text, at least. However why it not denoising when requirement for upscale is not met? It is something that heavily related on something or just synthetical limit?

**Comment by TianZerL** [comment]
>  it use CPU for denoising.
GLSL shader only runs in GPU.

**Comment by XCanG** [comment]
> > it use CPU for denoising.
> 
> GLSL shader only runs in GPU.

Hm, I only see 20% GPU usage and 3 cores to 100%.

**Comment by TianZerL** [comment]
> Hm, I only see 20% GPU usage and 3 cores to 100%.

What's your CPU?   

CPU only does scheduling tasks of shaders. The more shaders you load, the higher usage CPU get. Many layers of shaders are integrated in a single file of the UL shader.

**Comment by XCanG** [comment]
Ryzen 9 3900X

**Comment by bloc97** [comment]
Look forward to the v3.2 update, as it will incorporate a major overhaul to the denoising algorithms. However this update will be delayed as my personal training computer died and I won't use university resources for a personal extracurricular project. New GPUs are also very much inexistent right now...... I can't promise a date, but I think Q1 2021 would be reasonable.

**Comment by XCanG** [comment]
I'll set this repository for watch and will wait.

**Comment by Artins90** [comment]
> I'll set this repository for watch and will wait.

I did a quick test using the current 4.0 RC with the following shaders (my shader configuration might not be optimal): 
`glsl-shaders="~~/shaders/Anime4K_Restore_CNN_Light_M.glsl;~~/shaders/Anime4K_Upscale_Denoise_CNN_x2_L.glsl;~~/shaders/Anime4K_AutoDownscalePre_x4.glsl;~~/shaders/Anime4K_Restore_CNN_Moderate_L.glsl;~~/shaders/Anime4K_Upscale_Denoise_CNN_x2_M.glsl;"`

https://imgsli.com/NjUwNjU

At native 4K it looks great and the noise is basically gone when viewing in full-screen. 
Really looking forward to the final release and to further tweaks/options to the shader combinations.

**Comment by bloc97** [comment]
The new shaders are much simpler/easier to use. For 1080p->4K it is simply:
`~~/shaders/Anime4K_Clamp_Highlights.glsl;~~/shaders/Anime4K_Restore_CNN_Moderate_VL.glsl;~~/shaders/Anime4K_Upscale_CNN_x2_L.glsl`
Here is what I get:
https://imgsli.com/NjUwNzU

I have updated the optimized [GLSL instructions](https://github.com/bloc97/Anime4K/blob/master/GLSL_Instructions.md) for 4.x.

**Comment by Jules-A** [comment]
> The new shaders are much simpler/easier to use. For 1080p->4K it is simply:
> `~~/shaders/Anime4K_Clamp_Highlights.glsl;~~/shaders/Anime4K_Restore_CNN_Moderate_VL.glsl;~~/shaders/Anime4K_Upscale_CNN_x2_L.glsl`
> Here is what I get:
> https://imgsli.com/NjUwNzU
> 
> I have updated the optimized [GLSL instructions](https://github.com/bloc97/Anime4K/blob/master/GLSL_Instructions.md) for 4.x.

That doesn't really show the effectiveness of denoising though?
How is the speed vs the old Bilateral_Median? 

The new shaders are a little puzzling for me, they seem to reduce lines a bit too much for my liking even though I previously used line thinning with 0.36 Strength and 2 iterations along with slightly reduced strength line darkening. Do I forgo those shaders now and edit something in another file?

**Comment by bloc97** [comment]
I can try to explain what the shader does. The "thinning lines" effect is actually an illusion. The shader does not thin lines at all. It recovers original lines from badly resampled lines which happen often when digitally compositing anime.

Consider this test pattern, brush strokes of size 1 to 10 from left to right:
![strokes_test](https://user-images.githubusercontent.com/567732/128979664-bde6cfca-fbe4-4639-ab83-28dec4cd54ca.png)

We can artificially degrade it by downscaling the image by 40% (1080p->648p) using then upscale back to 1080p using bicubic.
Then we apply JPEG compression of 50%, which is similar to moderately compressed anime seen online (streams/youtube).
![strokes_test_lr](https://user-images.githubusercontent.com/567732/128980026-8b2c66ab-c189-48f0-9460-392cc675ef08.jpg)
If you look at the brush strokes of size 1 and 2, they look much larger now due to the nonlinear degradation we caused by resampling and compressing the image.

Then we pass it through the line reconstruction shader (moderate, VL variant):
![strokes_test_sr](https://user-images.githubusercontent.com/567732/128980229-bcf5f3ff-45b0-4ab9-addd-4ef2c1c6c5a0.png)

It now looks almost exactly like the original image, without noise and bicubic degradation.
Best viewed zoomed in, they are 1080p images.

**Comment by bloc97** [comment]
Another interesting experiment is to try on different amounts of degradation for the 1 and 2 width strokes.
HR Image:
![strokes_test_lr_variation](https://user-images.githubusercontent.com/567732/128982488-14de3686-2ffd-4a74-808e-beea450e39c9.png)

JPEG compression of 50%
![strokes_test_lr_variation](https://user-images.githubusercontent.com/567732/128982450-9cc28dec-36e4-45d0-8172-41d7af562d52.jpg)

After the line reconstruction shader (moderate, VL variant):
![strokes_test_variation_sr](https://user-images.githubusercontent.com/567732/128982642-7f9b699d-b61a-4d1b-8bf7-748591d32394.png)
Most of the lines have the same thickness, but 100% and 90% lines are slightly thinned as they are outliers. 

However, anime will almost never have these two cases except for text overlays, baked subtitles and high quality background images. (Try to find these edge cases in modern 1080p anime, I tried and could not... Could be proven wrong.) Furthermore, the edge cases are still very hard for the relatively small network to learn. You would need advanced architectures with millions of parameters, which makes real time out of the question.

**Comment by Jules-A** [comment]
That's pretty cool info but I don't really care too much about what the shader is really doing as long as the end result is good and I was really happy with how those shaders looked on 720/1080p decently encoded anime on anything that wasn't built-in text. It did however make low res/poorly encoded videos unwatchable though. 

Also... Umm... you didn't answer any of my questions :/

**Comment by bloc97** [comment]
@Jules-A currently the new shaders don't work well with older anime, I still need to collect more data. Do you know which anime are problematic? Some screenshots of original low res images would be very appreciated. Also most of the older shaders are still going to be supported except a few like deblur-cnn, you can still use them, is there one shader that you used that I removed?

The aliasing is still being worked on, if you have some screenshots that I can put in the validation set that would be of big help!

**Comment by Jules-A** [comment]
> 
> 
> @Jules-A currently the new shaders don't work well with older anime, I still need to collect more data. Do you know which anime are problematic? Some screenshots of original low res images would be very appreciated. Also most of the older shaders are still going to be supported except a few like deblur-cnn, you can still use them, is there one shader that you used that I removed?
> 
> The aliasing is still being worked on, if you have some screenshots of cases with severe aliasing (before applying shaders) that I can put in the validation set that would be of big help!

Pretty much any older anime with a 480p base resolution. I only have Honey and Clover (downloaded straight from CR) but here's a comparison of my current shaders: https://i.imgur.com/9FRbNR2.png   and some lowish settings in MadVR: https://i.imgur.com/PUNfkV9.png . I've had far worse problems with other old anime but I remember having major issues in Le Chevalier D'Eon from Funimation (the show also had a lot of noise that the billinear denoisers weren't good enough).

As for aliasing before shaders, I'm really not sure, they just seem to get severely aliased by the shaders rather than starting out aliased.

**Comment by Artins90** [comment]
The current version does add some slight softness to backgrounds and textures in modern 1080p anime (you might need a 4k screen to notice it or zooming).
For example the railing in this comparison is a bit blurrier than in the source: https://imgsli.com/NjUzMzU
Between the two, personally I still prefer Anime4k since in it's current state the pros far outweigh the small softness added to backgrounds.

**Comment by Jules-A** [comment]
> I though we were talking about the new line reconstruction shaders? 
> Some overshoot (over sharpening) can be seen but it can be fixed by using the soft versions of restore instead (in the repo, not yet released).

Oh sorry, I haven't had time to properly play around with the new shaders so I'm just using the old 3.1 shaders as known good for 720/1080p anime. I just quickly tested 4.0 and the aliasing is pretty noticeable (you can see it even in Artins comparison). It's probably not noticeable or annoying for most people but I'm very sensitive to aliasing for some reason.
